File: yizhi/LeavesHack-1.4.0/src/main/java/com/dev/leavesHack/utils/combat/CombatUtil.java
package com.dev.leavesHack.utils.combat;

import com.google.common.collect.Lists;
import meteordevelopment.meteorclient.MeteorClient;
import meteordevelopment.meteorclient.systems.friends.Friends;
import net.minecraft.client.network.AbstractClientPlayerEntity;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.PlayerEntity;

import java.util.ArrayList;
import java.util.List;

import static meteordevelopment.meteorclient.MeteorClient.mc;

public class CombatUtil {
    public static List<PlayerEntity> getEnemies(double range) {
        List<PlayerEntity> list = new ArrayList<>();
        for (AbstractClientPlayerEntity player : Lists.newArrayList(mc.world.getPlayers())) {
            if (!isValid(player, range)) continue;
            list.add(player);
        }
        return list;
    }
    public static boolean isValid(Entity entity, double range) {
        boolean invalid = entity == null || !entity.isAlive() || entity.equals(mc.player) || entity instanceof PlayerEntity player && Friends.get().isFriend(player) || mc.player.getPos().distanceTo(entity.getPos()) > range;

        return !invalid;
    }
    public static boolean isValid(Entity entity) {
        boolean invalid = entity == null || !entity.isAlive() || entity.equals(mc.player) || entity instanceof PlayerEntity player && Friends.get().isFriend(player);

        return !invalid;
    }
    public static PlayerEntity getClosestEnemy(double distance) {
        PlayerEntity closest = null;

        for (PlayerEntity player : getEnemies(distance)) {
            if (closest == null) {
                closest = player;
                continue;
            }

            if (!(mc.player.squaredDistanceTo(player.getPos()) < mc.player.squaredDistanceTo(closest))) continue;

            closest = player;
        }
        return closest;
    }
}

File: src/main/kotlin/com/zychen027/MeteorPlusPlus/utils/combat/CombatUtil.kt
package com.zychen027.meteorplusplus.utils.combat

import meteordevelopment.meteorclient.MeteorClient.mc
import meteordevelopment.meteorclient.systems.friends.Friends
import net.minecraft.entity.Entity
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.player.PlayerEntity

object CombatUtil {

    fun getEnemies(range: Double): List<PlayerEntity> {
        val list = mutableListOf<PlayerEntity>()
        for (player in mc.world?.players ?: emptyList()) {
            if (!isValid(player, range)) continue
            list.add(player)
        }
        return list
    }

    fun isValid(entity: Entity?, range: Double): Boolean {
        if (entity == null || !entity.isAlive) return false
        val player = mc.player ?: return false
        if (entity == player) return false
        if (entity is PlayerEntity && Friends.get().isFriend(entity)) return false
        return player.distanceTo(entity) <= range
    }

    fun isValid(entity: Entity?): Boolean {
        if (entity == null || !entity.isAlive) return false
        if (entity == mc.player) return false
        if (entity is PlayerEntity && Friends.get().isFriend(entity)) return false
        return true
    }

    fun getClosestEnemy(distance: Double): PlayerEntity? {
        var closest: PlayerEntity? = null
        for (player in getEnemies(distance)) {
            if (closest == null) {
                closest = player
                continue
            }
            val playerDist = mc.player?.squaredDistanceTo(player) ?: Double.MAX_VALUE
            val closestDist = mc.player?.squaredDistanceTo(closest) ?: Double.MAX_VALUE
            if (playerDist >= closestDist) continue
            closest = player
        }
        return closest
    }

    /**
     * 重置攻击冷却（适配 1.21.11 yarn 映射）
     *
     * 1.21.11 映射名：
     *   - lastAttackedTime  (field_6230)
     *   - ticksSinceLastAttack (field_6273)
     *
     * 没有 resetLastAttackedTicks() 方法，只能反射写字段
     */
    fun resetAttackCooldown(player: LivingEntity) {
        // 优先尝试 1.21.11 yarn: lastAttackedTime
        try {
            val field = LivingEntity::class.java.getDeclaredField("lastAttackedTime")
            field.isAccessible = true
            field.setInt(player, 0)
            return
        } catch (_: Exception) {}

        // 降级: ticksSinceLastAttack
        try {
            val field = LivingEntity::class.java.getDeclaredField("ticksSinceLastAttack")
            field.isAccessible = true
            field.setInt(player, 0)
            return
        } catch (_: Exception) {}

        // 再降级: 旧版映射名 lastAttackedTicks
        try {
            val field = LivingEntity::class.java.getDeclaredField("lastAttackedTicks")
            field.isAccessible = true
            field.setInt(player, 0)
        } catch (_: Exception) {}
    }
}
